src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -163,60 +156,187 @@ private void CheckSourceLocationSeen(AnonymousTypePublicSymbol anonymous)
+            return this.SynthesizedDelegates.GetOrAdd(key, synthesizedDelegate);
         }
 
         private NamedTypeSymbol ConstructAnonymousDelegateImplementationSymbol(AnonymousDelegatePublicSymbol anonymous, int generation)
         {
             var typeDescr = anonymous.TypeDescriptor;
             Debug.Assert(typeDescr.Location.IsInSource); // AnonymousDelegateTemplateSymbol requires a location in source for ordering.
 
-            var fields = typeDescr.Fields;
-            bool returnsVoid = fields.Last().Type.IsVoidType();
-            int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
-            var refKinds = default(RefKindVector);
-            if (fields.Any(f => f.RefKind != RefKind.None))
+            // If all parameter types and return type are valid type arguments, construct
+            // the delegate type from a generic template. Otherwise, use a non-generic template.
+            if (!allValidTypeArguments(typeDescr))
             {
-                refKinds = RefKindVector.Create(nTypeArguments);
+                var fields = typeDescr.Fields;
+                bool returnsVoid = fields.Last().Type.IsVoidType();
+                int nTypeArguments = fields.Length - (returnsVoid ? 1 : 0);
+                var refKinds = default(RefKindVector);
+                if (fields.Any(f => f.RefKind != RefKind.None))
+                {
+                    refKinds = RefKindVector.Create(nTypeArguments);
+                    for (int i = 0; i < nTypeArguments; i++)
+                    {
+                        refKinds[i] = fields[i].RefKind;
+                    }
+                }
+                var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(nTypeArguments);
                 for (int i = 0; i < nTypeArguments; i++)
                 {
-                    refKinds[i] = fields[i].RefKind;
+                    typeArgumentsBuilder.Add(fields[i].TypeWithAnnotations);
+                }
+                var typeArguments = typeArgumentsBuilder.ToImmutableAndFree();
+                var template = SynthesizeDelegate(parameterCount: fields.Length - 1, refKinds, returnsVoid, generation);
+
+                Debug.Assert(typeArguments.Length == template.TypeParameters.Length);
+                return typeArguments.Length == 0 ?
+                    template :
+                    template.Construct(typeArguments);
+            }
+            else
+            {
+                var typeParameters = GetReferencedTypeParameters(typeDescr);
+                var key = getTemplateKey(typeDescr, typeParameters);
+
+                // Get anonymous delegate template
+                AnonymousDelegateTemplateSymbol? template;
+                if (!this.SynthesizedDelegates.TryGetValue(key, out template))
+                {
+                    template = this.SynthesizedDelegates.GetOrAdd(key, new AnonymousDelegateTemplateSymbol(this, typeDescr, typeParameters));
```

Comment:
>new AnonymousDelegateTemplateSymbol(this, typeDescr, typeParameters) [](http://example.com/codeflow?start=71&length=68)

It feels like the logic could be smarter and the fields passing the ```isValidTypeArgument``` check could be lifted to additional type parameters. I.e. the only non-type-parameters types in the signature would be those that couldn't be used as type arguments.

Reply:
We can add that optimization later if needed.

Reply:
> We can add that optimization later if needed. 

Would you mind opening an issue for triage?

Reply:
[https://github.com/dotnet/roslyn/issues/58537](https://github.com/dotnet/roslyn/issues/58537)

Reply:
Thank you.